RUFAS/input_manager.py

```diff
@@ -53,120 +76,285 @@ def _load_metadata(self, metadata_path: str = "input/example_metadata.json") ->
-                                   info_map)
-            except Exception as e:
-                raise e
+        om.add_log("open_csv_file", f"Attempting to open {file_path}.", info_map)
+        try:
+            with open(file_path, "r") as csv_file:
+                data_frame = pd.read_csv(csv_file)
+                data_dict = {column: data_frame[column].tolist() for column in data_frame.columns}
+                if not data_frame.empty:
+                    om.add_log("load_data_successful",
+                               f"Successfully loaded data from {file_path}.",
+                               info_map)
+                return data_dict
+        except Exception as e:
+            raise e
 
-    def _validate_data(self, eager_termination: bool = True) -> bool:
+    def _validate_data(self, eager_termination: bool) -> bool:
         """
-        Validates input data and attempts to fix any invalid input data.
+        Validates data from input files and adds data to pool if valid.
 
         Parameters
         ----------
-        eager_termination : bool, default=True
-            If true, the process will be terminated upon finding invalid data.
+        eager_termination : bool
+            If True, the process will be terminated as soon as finding invalid data and failing to fix it.
+            If False, the process will be terminated after going through and validating the entire data,
+            if invalid data is found.
 
         Returns
         -------
         bool
-            True if all data is valid; False otherwise.
+            True if data is valid, otherwise False.
         """
         info_map = {"class": self.__class__.__name__,
                     "function": self._validate_data.__name__,
                     }
-        valid_elements_counter = 0
-        invalid_elements_counter = 0
-        fixed_elements_counter = 0
-        invalid_critical_elements_counter = 0
-        total_elements_checked_counter = 0
-
-        for key in self.__pool.keys():
-            for variable, value in self.__pool[key].items():
-                total_elements_checked_counter += 1
-                if self._validate_element(variable, value):
-                    valid_elements_counter += 1
+        valid_items_counter = 0
+        invalid_critical_items_counter = 0
+        total_items_counter = 0
+
+        for module_key, details in self.__metadata["files"].items():
+            file_path = details["path"]
+            if details["type"] == "json":
+                data = self._load_data_from_json(file_path)
+            elif details["type"] == "csv":
+                data = self._load_data_from_csv(file_path)
+            else:
+                om.add_warning("InputManager load data file is not csv/json",
+                               f"{module_key} data must be available in either csv or json file type.",
+                               info_map)
+                continue
```

Comment:
As discussed on July 17th, avoid using `elif` structure for routing the flow to functions. 
```suggestion
            data_type_to_loadder_map = {
                "json": self._load_data_from_json(file_path),
                 "csv": self._load_data_from_csv(file_path),
            }
            try:
                data_type_to_loadder_map.get(details["type"])
            except:
               raise KeyException(f"Faulty data type in {module_key}, only CSV and JSON are supported."}
```

Reply:
In conversations and testing with @michaelRichards99, instantiating a dictionary with multiple function calls as values will cause all functions to be called at instantiation (in this case with the `file_path` argument). This will lead to a number of unintended consequences so this isn't a good approach.
![image (1)](https://github.com/RuminantFarmSystems/MASM/assets/70217952/0d03a5d7-7a0a-49eb-b648-270fd07bf86c)
![image (2)](https://github.com/RuminantFarmSystems/MASM/assets/70217952/dc3a78a0-8800-47ce-902b-19cf1982c738)

Reply:
simply move the arg from lines 162 and 163 to line 166.
Running 
```
print("Start")
class foo():
    def __init__(self):
        print("init, creating func mapping dict")
        d={
            "bar1": self.bar1,
            "bar2": self.bar2,
            "bar3": self.bar3,
            "bar4": self.bar4,
        }
        print("dict is created")
        x=1
        for k,v in d.items():
            if k=="bar2":
                continue
            print(f"calling {k}")
            v(x)
            x+=1
    def bar1(self, x):
        print(f"bar1 {x=}")
    def bar2(self, x):
        print(f"bar2 {x=}")
    def bar3(self, x):
        print(f"bar3 {x=}")
    def bar4(self, x):
        print(f"bar4 {x=}")
print("creating foo object")
f=foo()
```
Generates
```
Start
creating foo object
init, creating func mapping dict
dict is created
calling bar1
bar1 x=1
calling bar3
bar3 x=2
calling bar4
bar4 x=3
```
As you see, bar2 is not called.
CC @michaelRichards99

Reply:
That works cleanly in this instance because the function's arity is the same.

It does become more difficult for variadic functions as is the case for the 2nd instance of this in this PR. In that case, the types `string`, `array`, and `number` all have functions with 3 parameters and the `bool` type needs a function that always returns true. @ew3361zh An additional `self._validate_bool_type_element(variable_properties, var_name, input_data_value)` that always returns `True` would remove this complexity and then the dictionary that stores function pointers works as planned.

Reply:
Can you write a piece of code that shows your point?

Reply:
We had the following dictionary. (I've taken out the parameters so it doesn't call the functions and replaced True with bool().
```
func_map = {
    "string": self._validate_string_type_element,
    "number": self._validate_num_type_element,
    "array": self._validate_array_type_element,
    "bool": bool
}
```
bool() takes 1 arg while the others have 3 (variable_properties, var_name, input_data_value). when you get the function with `func_map(var_type)`, you'd need to check if the type was a `bool` so you can pass the proper arguments to avoid a TypeError.

Instead of special-casing the `bool` type, we could introduce `self._validate_bool_type_element(variable_properties, var_name, input_data_value)` which removes the complexity.

Reply:
Yes, bool can have a function. It will do nothing right now but allows adding validation for bool down the road.

Reply:
Updated.